File: algorithm study/BOJ/3. Gold/21758.js
// boj 21758 꿀 따기
// prefix sum

const input = require("fs").readFileSync("./example.txt").toString().split("\n");
// const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

let n = input[0].split(" ").map(Number); //3~100,000

let arr = [];
arr = input[1].split(" ").map(Number);
arr.splice(0, 0, 0);

let temp = []; //누적합
temp[0] = arr[0];
for (let i = 1; i <= n; i++) temp[i] = temp[i - 1] + arr[i];
// console.log(arr);
// console.log(temp);

let maximum = -1;
//1.꿀통-벌-벌: 꿀은 1, 오른쪽 벌은 n -> 가운데 벌을 i로 돌림
for (let i = 2; i < n; i++) {
  let honey = temp[n - 1] - arr[i] + temp[i - 1];
  maximum = Math.max(maximum, honey);
}

//2. 벌-꿀통-벌: 꿀을 i로 돌림
for (let i = 2; i < n; i++) {
  let honey = temp[i] - arr[1] + (temp[n - 1] - temp[i - 1]);
  maximum = Math.max(maximum, honey);
}

//3. 벌-벌-꿀통: 왼쪽 벌은 1, 꿀은 n -> 가운데 벌을 i로 돌림
for (let i = 2; i < n; i++) {
  let honey = temp[n] - arr[1] - arr[i] + (temp[n] - temp[i]);
  maximum = Math.max(maximum, honey);
}
console.log(maximum);

File: algorithm study/BOJ/3. Gold/21758.cpp
// boj 21758 꿀 따기
// prefix sum
#include <iostream>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int t;
    cin >> t;
    int* arr = new int[t + 1]{0};
    for (int i = 1; i <= t; i++) cin >> arr[i];
    // for (int i = 0; i <= t; i++) cout << arr[i] << " ";

    int* prefixsum = new int[t + 1]{0};
    prefixsum[1] = arr[1];
    for (int i = 1; i <= t; i++) prefixsum[i] = prefixsum[i - 1] + arr[i];
    // for (int i = 0; i <= t; i++) cout << prefixsum[i] << " ";

    int maximum = -1;
    // 1.꿀통-벌-벌: 꿀은 1, 오른쪽 벌은 n -> 가운데 벌을 i로 돌림
    for (int i = 2; i < t; i++) {
        int honey = (prefixsum[t - 1] - arr[i]) + prefixsum[i - 1];
        maximum = max(honey, maximum);
    }

    // 2. 벌-꿀통-벌: 꿀을 i로 돌림
    for (int i = 2; i < t; i++) {
        int honey = (prefixsum[i] - arr[1]) + (prefixsum[t - 1] - prefixsum[i - 1]);
        maximum = max(honey, maximum);
    }

    // 3. 벌-벌-꿀통: 왼쪽 벌은 1, 꿀은 n -> 가운데 벌을 i로 돌림
    for (int i = 2; i < t; i++) {
        int honey = (prefixsum[t] - arr[1] - arr[i]) + (prefixsum[t] - prefixsum[i]);
        maximum = max(honey, maximum);
    }

    cout << maximum;
    return 0;
}
